saleor/attribute/migrations/0020_auto_20220214_1027.py

```diff
@@ -0,0 +1,122 @@
+# Generated by Django 3.2.12 on 2022-02-14 10:27
+
+from collections import defaultdict
+
+from django.db import migrations
+from django.db.models import Exists, OuterRef
+
+
+def fetch_reference_id_from_slug(slug):
+    ref_id = slug.split("_")[1]
+    return int(ref_id)
+
+
+def import_attribute_values(attr_values_qs, Model):
+    attr_values_to_update = []
+    attr_values_to_delete = set()
+
+    attr_value_map = defaultdict(list)
+    for attr_value in attr_values_qs:
+
+        ref_id = fetch_reference_id_from_slug(attr_value.slug)
+        attr_value_map[ref_id].append(attr_value)
```

Comment:
In your solution, you iterate through all `AttributeValues` two times. I will suggest firstly getting all attribute values slugs with the use of `values_list`, then preparing a dict with the attribute value `slug` as a key and `ref_id` as a value. In the next step, we should get all reference objects with the use of the `in_bulk` method. Next, we can iterate through every value and check if the given attribute value is in our slug - reference mapping. If it is, we should assign reference object to `reference_page` or `reference_product` field and add this attr value to `attr_values_to_update` list (it should be also set by the way). Otherwise, add  attr value to `attr_values_to_delete` set.